### PR TITLE
MWPW-109329 - Placeholders

### PR DESCRIPTION
**Pull request by auniverseaway**
* Add support for placeholders via regex
* Add support for placeholders via key
* Update library to support placeholders
* Improve library AX

Resolves: [MWPW-109329](https://jira.corp.adobe.com/browse/MWPW-109329)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/tools/library
- After: https://placeholders--milo--adobecom.hlx.page/tools/library

### CODE REVIEW

File: libs/blocks/library-config/lists/placeholders.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { createTag } from '../../../utils/utils.js';
+import createCopy from '../library-utils.js';
+
+async function fetchPlaceholders(path) {
+  const resp = await fetch(path);
+  if (!resp.ok) return;
+  const json = await resp.json();
+  if (!json.data) return;
+  return json.data;
```
Comment (chrischrischris) [inline libs/blocks/library-config/lists/placeholders.js:9]: total edge case, but if returning `undefined` then L14 will fail.
```suggestion
  if (!resp.ok) return [];
  const json = await resp.json();
  return json.data || [];
```

File: libs/blocks/library-config/lists/blocks.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    title.addEventListener('click', () => {
+      title.classList.toggle('is-open');
+    });
+
+    previewButton.addEventListener('click', (e) => {
+      e.stopPropagation();
+      window.open(block.path, '_blockpreview');
+    });
+
+    // eslint-disable-next-line no-await-in-loop
```
Comment (chrischrischris) [inline libs/blocks/library-config/lists/blocks.js:73]: I think this and L77 can be deleted

File: libs/blocks/library-config/library-config.js [hunk 3]
```diff
@@ -1,192 +1,137 @@
...
+  });
+  return json;
+}
 
-  el.querySelector('div').remove();
+async function getSuppliedLibrary() {
+  const url = new URL(window.location.href);
+  const library = url.searchParams.get('library');
+  if (!library) return null;
+  return await fetchLibrary(`https://main--${library}.hlx.live`);
```
Comment (chrischrischris) [inline libs/blocks/library-config/library-config.js:45]: Add support for `passConfig` (https://github.com/adobe/helix-sidekick-extension/pull/168)

Also I _think_ that the `await` can be removed when returning.

```suggestion
  const library = url.searchParams.get('library');
  if (library) return await fetchLibrary(`https://main--${library}.hlx.live`);
  const repo = url.searchParams.get('repo');
  const owner = url.searchParams.get('owner');
  if (repo && owner) return await fetchLibrary(`https://main--${repo}--${owner}.hlx.live`
```

File: libs/blocks/library-config/library-config.js [hunk 4]
```diff
@@ -1,192 +1,137 @@
...
-    });
-  blockPaths.push(...additionalBlockLibs);
-
-  const libraries = [ { text: 'Blocks', paths: blockPaths } ];
+async function fetchLibrary(domain) {
+  const resp = await fetch(`${domain}${LIBRARY_PATH}`);
+  if (!resp.ok) return;
+  const json = await resp.json();
+  Object.keys(json).forEach((key) => {
+    if (key.startsWith(':')) delete json[key];
```
[the remark this answers is not in the document]
Comment (auniverseaway) [reply libs/blocks/library-config/library-config.js:36]: Originally I wanted to just iterate, but we're explicitly now opting into certain library types. We'll eventually need this if we want to be more generic, but I'll remove for now.

[the remark this answers is not in the document]
Comment (auniverseaway) [reply libs/blocks/library-config/library-config.js]: Resolved.

File: libs/blocks/library-config/library-config.js (hunk 3 above)
Comment (auniverseaway) [reply libs/blocks/library-config/library-config.js]: I went with...

```js
async function getSuppliedLibrary() {
  const { searchParams } = new URL(window.location.href);
  const repo = searchParams.get('repo');
  const owner = searchParams.get('owner');
  if (!repo && !owner) return null;
  return fetchLibrary(`https://main--${repo}--${owner}.hlx.live`);
}
```

**Review by meganthecoder — APPROVED** [review]

**Pull request merged**